lib/msf/core/exploit/remote/kerberos/service_authenticator/base.rb

```diff
@@ -193,35 +221,108 @@ def authenticate(options = {})
+    
+    if mechanism == 'spnego'
+      security_blob = encode_gss_spnego_ap_request(ap_request_asn1)
+    elsif mechanism == 'kerberos'
+      security_blob = encode_gss_kerberos_ap_request(ap_request_asn1)
+    end
 
     {
-      security_blob: encode_gss_ap_request(ap_request_asn1),
-      session_key: tgs_auth.key
+      security_blob: security_blob,
+      session_key: tgs_auth.key,
+      client_sequence_number: sequence_number
     }
   end
 
+  def get_message_encryptor(key, client_sequence_number, server_sequence_number)
+    Rex::Proto::Gss::Kerberos::MessageEncryptor.new(key, 
+                                            client_sequence_number, 
+                                            server_sequence_number, 
+                                            true, # is_initiator
+                                            use_acceptor_subkey: true)
+  end
+
+  def parse_gss_init_response(token, session_key, mechanism: 'kerberos')
+    mech_id, encapsulated_token = unwrap_pseudo_asn1(token)
+
+    if mech_id.value == Rex::Proto::Gss::OID_KERBEROS_5.value
+      tok_id = encapsulated_token[0,2]
+      data = encapsulated_token[2, encapsulated_token.length]
+      case tok_id
+      when "\x02\x00" # AP_REP
+        ap_req = Rex::Proto::Kerberos::Model::ApRep.decode(data)
+        print_good("#{peer} - Received AP-REQ. Extracting session key...")
+
+        raise ::Rex::Proto::Kerberos::Model::Error::KerberosError, 'Mismatching etypes' if session_key.type != ap_req.enc_part.etype
+
+        decrypted = ap_req.decrypt_enc_part(session_key.value)
+        
+        result = {
+          :ap_rep_subkey => decrypted.subkey,
+          :server_sequence_number => decrypted.sequence_number
+        }
+      when "\x03\x00" # KRB_ERR
+        krb_err = Rex::Proto::Kerberos::Model::KrbError.decode(data)
+        print_error("#{peer} - Received KRB-ERR.")
+
+        raise ::Rex::Proto::Kerberos::Model::Error::KerberosError.new(res: krb_err)
+      else
+        raise ::Rex::Proto::Kerberos::Model::Error::KerberosError, "Unknown token id: #{tok_id.inspect}"
+      end
+    else
+      raise ::NotImplementedError, "Parsing mechtype #{mech_id.value} not supported"
+    end
+
+  end
+
+  def build_gss_ap_req_checksum_value(mutual_auth)
+    # No channel binding
+    channel_binding_info = "\x00" * 16
+    channel_binding_info_len = [channel_binding_info.length].pack('V')
+
+    flags = GSS_REPLAY_DETECT | GSS_SEQUENCE | GSS_CONFIDENTIAL | GSS_INTEGRITY
+    flags |= GSS_MUTUAL if mutual_auth
+
+    flags = [flags].pack('V')
+
+    checksum_val = channel_binding_info_len + channel_binding_info + flags
+
+    checksum = Rex::Proto::Kerberos::Model::Checksum.new(
+      type: Rex::Proto::Gss::KRB_AP_REQ_CHKSUM_TYPE,
+      checksum: checksum_val)
+  end
+
   def build_service_ap_request(opts = {})
     authenticator = opts.fetch(:authenticator) do
       build_authenticator(opts.merge(
         subkey: nil,
```

Comment:
@smashery Just confirming, were we expecting a subkey to be generated here?

Reply:
So I don't believe it's required to supply a subkey; in its absence, the TGS-negotiated session key is used (although ultimately, the key provided in the AP-REP is used).